### ISSUE TITLE
Would implement command INSERT INTO SELECT FROM ?

### ISSUE THREAD

**Issue by Jonwei** [Tencent/wcdb#436]
### The language of WCDB
Objective-C, Swift or Java 

### The version of WCDB
v1.0.7

### The platform of WCDB
iOS

### The installation of WCDB
Cocoapods

### What's the issue?

Would implement command INSERT INTO SELECT FROM , for table content copy as sqlite migration ?

SQLite office doc [Making Other Kinds Of Table Schema Changes](https://www.sqlite.org/lang_altertable.html)

**Comment by RingoD** [comment]
It's already supported since the first release.

**Comment by RingoD** [comment]
Here is the incomplete sample code `StatementInsert().insert(table, columns).values(StatementSelect().XXX)`. Check WINQ in the wiki and `values` interface in `StatementInsert`.

**Comment by Jonwei** [comment]
> Here is the incomplete sample code `StatementInsert().insert(table, columns).values(StatementSelect().XXX)`. Check WINQ in the wiki and `values` interface in `StatementInsert`.

Ok。 got it.  I'll try it.
thank you for response

**Comment by Jonwei** [comment]
@RingoD have two run error, please  help me  fix it. 
`
WCDB::ColumnList list = {
        WCDB::Column("demoID"),
        WCDB::Column("is_public"),
        WCDB::Column("desc")
    };
    WCDB::ColumnResultList resultList = {
        WCDB::ColumnResult(WCDB::Column("demoID")),
        WCDB::ColumnResult(WCDB::Column("is_public")),
        WCDB::ColumnResult(WCDB::Column("desc"))
    };
    WCDB::ExprList exprList = {WCDB::StatementSelect().select(resultList).from("JWDemo")};
    result = [self.database exec:(WCDB::StatementInsert().insert("new_JWDemo", list).values(exprList))];
    
//    result = [self.database dropTableOfName:@"JWDemo"];
    result = [self.database exec:(WCDB::StatementDropTable().drop("JWDemo"))];
`
===>>
[WCDB][DEBUG]Code:1, Type:SQLiteGlobal, Msg:near "SELECT": syntax error
[WCDB][ERROR]Code:1, Type:SQLite, Tag:0, Op:4, ExtCode:1, Msg:near "SELECT": syntax error, SQL:INSERT OR REPLACE INTO new_JWDemo(demoID,is_public,desc) VALUES(SELECT demoID,is_public,desc FROM JWDemo),

[WCDB][DEBUG]Code:6, Type:SQLiteGlobal, Msg:statement aborts at 13: [DROP TABLE IF EXISTS JWDemo] database table is locked
[WCDB][ERROR]Code:6, Type:SQLite, Tag:0, Op:4, ExtCode:6, Msg:database table is locked, SQL:DROP TABLE IF EXISTS JWDemo,

**Comment by RingoD** [comment]
OK, you find a bug!
It's our fault that do not have enough test case.
But it' fixed and redesigned(I think you already find that it not easy to use the WINQ syntax) in next version. With tests covered of course.
We will release the new version as fast as we can.

**Comment by Jonwei** [comment]
OK.  we figured out one error. 

Did you check `dropTableOfName:` method as above mentioned ?  
Below code-1 is my full code using `dropTableOfName:`, and it tells a error: database table is locked。
And, when change codes like code-2, it's OK, no error occurs. 
please help me fix it. 

【code-1】
`
- (BOOL)schemaVersion2ChangeTemp {
    BOOL result = false;
   
    // create new table
    result = [self.database createTableAndIndexesOfName:@"new_JWDemo" withClass:NSClassFromString(@"JWDemo")];
    NSLog(@"%s schema version: %lu", __PRETTY_FUNCTION__, (unsigned long)[self schemaVersion:NULL]);
    if (!result) {
        return result;
    }
    
   // get old table rows and insert to new table
    NSArray *objects = [self.database getAllObjectsOfClass:NSClassFromString(@"JWDemo") fromTable:@"JWDemo"];
    if (objects.count > 0) {
        result = [self.database insertObjects:objects into:NSClassFromString(@"new_JWDemo")];
        if (!result) {
            return result;
        }
    } else {
        //
    }
    
   // drop old table
    result = [self.database dropTableOfName:@"JWDemo"];
    NSLog(@"%s schema version: %lu", __PRETTY_FUNCTION__, (unsigned long)[self schemaVersion:NULL]);
    if (!result) {
        return result;
    }
    
    // rename new table name
    result = [self.database exec:(WCDB::StatementAlterTable().alter("new_JWDemo").rename("JWDemo"))];
    NSLog(@"%s schema version: %lu", __PRETTY_FUNCTION__, (unsigned long)[self schemaVersion:NULL]);
    if (!result) {
        return result;
    }
    
    return result;
}
`
===>>
[WCDB][DEBUG]Code:6, Type:SQLiteGlobal, Msg:statement aborts at 13: [DROP TABLE IF EXISTS JWDemo] database table is locked
[WCDB][ERROR]Code:6, Type:SQLite, Tag:0, Op:4, ExtCode:6, Msg:database table is locked, SQL:DROP TABLE IF EXISTS JWDemo,

【code-2】
`
- (BOOL)schemaVersion2ChangeTemp {
    BOOL result = false;
    
   // get old table rows and drop old table
    NSArray *objects = [self.database getAllObjectsOfClass:NSClassFromString(@"JWDemo") fromTable:@"JWDemo"];
    result = [self.database dropTableOfName:@"JWDemo"];
    NSLog(@"%s schema version: %lu", __PRETTY_FUNCTION__, (unsigned long)[self schemaVersion:NULL]);
    if (!result) {
        return result;
    }
    
    // create new table
    result = [self.database createTableAndIndexesOfName:@"new_JWDemo" withClass:NSClassFromString(@"JWDemo")];
    NSLog(@"%s schema version: %lu", __PRETTY_FUNCTION__, (unsigned long)[self schemaVersion:NULL]);
    if (!result) {
        return result;
    }
    
    // insert old table rows to  new table
    if (objects.count > 0) {
        result = [self.database insertObjects:objects into:NSClassFromString(@"new_JWDemo")];
        if (!result) {
            return result;
        }
    } else {
        //
    }
    
    // rename new table name
    result = [self.database exec:(WCDB::StatementAlterTable().alter("new_JWDemo").rename("JWDemo"))];
    NSLog(@"%s schema version: %lu", __PRETTY_FUNCTION__, (unsigned long)[self schemaVersion:NULL]);
    if (!result) {
        return result;
    }
    
    return result;
}
`

**Comment by RingoD** [comment]
It seems pretty good for both of these two methods.
Is it reproducible? Could you please try to reproduce it in a separated project and submit to us?

**Comment by Jonwei** [comment]
yes. it's reproducible. 
how do i send demo project to you ?

**Comment by Jonwei** [comment]
https://github.com/Jonwei/wcdbMigrationDemo  see this demo project

**Comment by RingoD** [comment]
@Jonwei  Hey, I opened your project and every things seems to be fine. I don't understand what you expect to show.

**Comment by RingoD** [comment]
First run:

> 2018-12-13 15:36:27.406936+0800 WCDBMigrationDemo[93303:804478] -[ViewController viewDidLoad] cachePath:
 /Users/sanhuazhang/Library/Developer/CoreSimulator/Devices/8F7F7D05-B58D-477B-BA72-375AFAB15787/data/Containers/Data/Application/FDD8501C-A04A-49F8-A83E-42A50A0C2B91/Documents
2018-12-13 15:36:46.988885+0800 WCDBMigrationDemo[93303:804478] -[DatabaseManager database] path: /Users/sanhuazhang/Library/Developer/CoreSimulator/Devices/8F7F7D05-B58D-477B-BA72-375AFAB15787/data/Containers/Data/Application/FDD8501C-A04A-49F8-A83E-42A50A0C2B91/Documents/demoMain.db
2018-12-13 15:36:47.000925+0800 WCDBMigrationDemo[93303:804478] -[DatabaseManager schemaVersion1Change] schema version: 1
2018-12-13 15:36:47.005485+0800 WCDBMigrationDemo[93303:804478] -[DatabaseManager migrate] result: 1

Changing ORM to:
```
//WCDB_PRIMARY(JWDemo, demoID)
WCDB_MULTI_PRIMARY(JWDemo, "MultiPrimaryConstraints", demoID)
WCDB_MULTI_PRIMARY(JWDemo, "MultiPrimaryConstraints", is_public)
```

Then run:

> 2018-12-13 15:37:02.888450+0800 WCDBMigrationDemo[93403:808300] -[ViewController viewDidLoad] cachePath:
 /Users/sanhuazhang/Library/Developer/CoreSimulator/Devices/8F7F7D05-B58D-477B-BA72-375AFAB15787/data/Containers/Data/Application/5C665D69-1236-496D-87EA-2DA5D1793921/Documents
2018-12-13 15:37:04.269245+0800 WCDBMigrationDemo[93403:808300] -[DatabaseManager database] path: /Users/sanhuazhang/Library/Developer/CoreSimulator/Devices/8F7F7D05-B58D-477B-BA72-375AFAB15787/data/Containers/Data/Application/5C665D69-1236-496D-87EA-2DA5D1793921/Documents/demoMain.db
[WCDB][DEBUG]Code:283, Type:SQLiteGlobal, Msg:recovered 3 frames from WAL file /Users/sanhuazhang/Library/Developer/CoreSimulator/Devices/8F7F7D05-B58D-477B-BA72-375AFAB15787/data/Containers/Data/Application/5C665D69-1236-496D-87EA-2DA5D1793921/Documents/
2018-12-13 15:37:04.274073+0800 WCDBMigrationDemo[93403:808300] -[DatabaseManager schemaVersion2Change] schema version: 2
2018-12-13 15:37:04.274403+0800 WCDBMigrationDemo[93403:808300] -[DatabaseManager schemaVersion2Change] schema version: 3
2018-12-13 15:37:04.274649+0800 WCDBMigrationDemo[93403:808300] -[DatabaseManager schemaVersion2Change] schema version: 4
2018-12-13 15:37:04.274923+0800 WCDBMigrationDemo[93403:808300] -[DatabaseManager migrate2] result: 1

It seems to be fine, too.

**Comment by Jonwei** [comment]
use migrateFail branch ?

**Comment by RingoD** [comment]
A prepared `WCTStatement` is not finalized in `schemaVersion:` method.

**Comment by RingoD** [comment]
It's fixed and tested in v1.1.0, which is the next version.
